src/frequenz/sdk/timeseries/_formula_engine/_formula_printer.py

```diff
@@ -0,0 +1,220 @@
+        """
+        return self.value > other.value
+
+
+    def __lt__(self, other: OperatorPrecedence) -> bool:
+        return self.value < other.value
+
+    def __le__(self, other: OperatorPrecedence) -> bool:
+        return self.value <= other.value
+
+    @staticmethod
+    def map(operator: str) -> OperatorPrecedence:
+        """Return the precedence of the operator.
+
+        Args:
+            operator: The operator.
+
+        Returns:
+            int: The precedence of the operator.
+        """
+        match operator:
+            case "+":
+                return OperatorPrecedence.Addition
+            case "-":
+                return OperatorPrecedence.Subtraction
+            case "*":
+                return OperatorPrecedence.Multiplication
+            case "/":
+                return OperatorPrecedence.Division
+            case _:
+                return OperatorPrecedence.Primary
+
+
+class StackItem:
+    """Represents a stack item for the formula printer."""
+
+    def __init__(self, value: str, precedence: OperatorPrecedence):
+        """Initialize the StackItem.
+
+        Args:
+            value: The value of the stack item.
+            precedence: The precedence of the stack item.
+        """
+        self.value = value
+        self.precedence = precedence
+
+    def __str__(self) -> str:
+        """Return the string representation of the stack item.
+
+        This is used for debugging purposes.
+
+        Returns:
+            str: The string representation of the stack item.
+        """
+        return f"(\"{self.value}\", {self.precedence})"
+
+    def plvalue(self, outer_precedence: OperatorPrecedence) -> str:
+        """Return the value of the stack item with parentheses if necessary.
+
+        Args:
+            outer_precedence: The precedence of the outer stack item.
+
+        Returns:
+            str: The value of the stack item with parentheses if necessary.
+        """
+        return f"({self.value})" if self.precedence < outer_precedence else self.value
+
+    def prvalue(self, outer_precedence: OperatorPrecedence) -> str:
+        """Return the value of the stack item with parentheses if necessary.
+
+        Args:
+            outer_precedence: The precedence of the outer stack item.
+
+        Returns:
+            str: The value of the stack item with parentheses if necessary.
+        """
+        return f"({self.value})" if self.precedence < outer_precedence else self.value
+
+    @staticmethod
+    def create_binary(lhs: StackItem, op: str, rhs: StackItem) -> StackItem:
+        """Create a binary stack item.
+
+        Args:
+            lhs: The left-hand side of the binary operation.
+            op: The operator of the binary operation.
+            rhs: The right-hand side of the binary operation.
+
+        Returns:
+            StackItem: The binary stack item.
+        """
+        pred = OperatorPrecedence.map(op)
```

Comment:
If `map` is only used here, where I assume you know already this is a binary op, then another alternative could just be:

```suggestion
        pred = OperatorPrecedence(op)
```

And you can remove `map()` completely.

Reply:
@llucax I've done that. But all I won was getting rid of `map` (which I renamed to `from_str`) but I had to add `__str__` instead, because I needed to convert back to a string representation, as `LHS op RHS` I need to print a proper string for `op` :).

Reply:
For for the fun of it, a shorter (but slightly slower) implementation. Too bad that the sorting of the string doesn't match the sort we need :P

```python
import functools

class OperatorPrecedence(enum.Enum):
    """The precedence of an operator."""

    SUBTRACTION = "-"
    ADDITION = "+"
    DIVISION = "/"
    MULTIPLICATION = "*"
    PRIMARY = "<primary>"

    @functools.cache
    def __lt__(self, other: OperatorPrecedence) -> bool:
        """Test the precedence of this operator is less than the precedence of the other operator.

        Args:
            other: The other operator (on the right-hand side).

        Returns:
            Whether the precedence of this operator is less than the other operator.
        """
        precedence = list(OperatorPrecedence)
        return precedence.index(self) < precedence.index(other)
```